### ISSUE TITLE
perf: table scan with too many delete tombstones is slow

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#12903]
> If I understand it correctly, all full scans on such tables will be affected, right? If we make specific fixes or improvements to Backfill, users may still find it confusing that batch scanning a temporal-filtered table is slow. This could be another reason to consider fixing it in a more general way...

_Originally posted by @BugenZhao in https://github.com/risingwavelabs/risingwave/issues/12680#issuecomment-1755619597_

**Comment by kwannoel** [comment]
Is this generalizable to some sort of cache? Such that we can handle tombstone ranges.
> Maintain table statistics to guide backfill. A simple statistic is per table min/max key. Backfill can use vnode | min_key as the start_key when it begins. This can solve a subset of the issues when the upstream table performs sequential deletes (e.g. TTL).

> Tune compaction to be more aggressive when the SST delete ratio is high. This can improve the situation but cannot solve it completely since it order to compaction to clean up a tombstone key completely, a full compaction is needed. When the data volume is high, full compaction may not be done in a fast manner even though compaction is triggered aggressively.
I think @Li0k has worked on this.

**Comment by Li0k** [comment]
> > Tune compaction to be more aggressive when the SST delete ratio is high. This can improve the situation but cannot solve it completely since it order to compaction to clean up a tombstone key completely, a full compaction is needed. When the data volume is high, full compaction may not be done in a fast manner even though compaction is triggered aggressively.
> 
> Can I confirm again @Li0k? This is done by your PRs right?

yeah, but cannot solve it completely, If the tombstone is affecting performance, we can consider adjusting the frequency of the tombstone picker triggers

**Comment by kwannoel** [comment]
> since it order to compaction to clean up a tombstone key completely, a full compaction is needed. When the data volume is high, full compaction may not be done in a fast manner even though compaction is triggered aggressively.

Further discussed offline with @Li0k , this part is still under discussion.

> Tune compaction to be more aggressive when the SST delete ratio is high. 

This part is done, compaction will occur in 10 min intervals. See https://github.com/risingwavelabs/risingwave/pull/12776.

**Comment by kwannoel** [comment]
For testing compaction: https://github.com/risingwavelabs/kube-bench/issues/306
